### PR TITLE
post user story banco do brasil

### PR DESCRIPTION
**Pull request by felipewind**
User story from Banco do Brasil.
This post was approved by the company.
Please tell me if any change is necessary.

### CODE REVIEW

**Review by gastaldi — COMMENTED** [review]
Spotted a few typos, looks good overall

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+* *Daily* - on every work day.
+* *On demand* - when the data is requested by the financial specialists or directly by the customer.
+
+
+== Technical Challenges to extract data from Open Banking
+There were some difficulties in accomplishing this task.
+
+=== Number of clients to extract and the API throttling
+
+We have millions of customers, so the application must be able to scale up to handle all the processing. At first there were not so many clients, but this number is growing daily.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:50]:
```suggestion
We have millions of customers, so the application must be able to scale up to handle all the processing. At first there were not so many clients using this feature, but this number is growing daily.
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+The difficulty is that sometimes some financial institutions have problems, so it is not possible to be sure that the Open Banking information will be available every day. 
+
+It is necessary to create some rules, where we can reprocess information from previous days.
+
+=== Observability
+
+We need to have some metrics to be able to know if the system is working correctly or if there is a problem. 
+
+For instance, "why is the data extraction is taking so long today?" "Is it some financial institution that is having a problem?" "Is it our system?" "Our database?"
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:72]:
```suggestion
For instance, "why is the data extraction taking so long today?" "Is it some financial institution that is having a problem?" "Is it our system?" "Our database?" 
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+It is necessary to create some rules, where we can reprocess information from previous days.
+
+=== Observability
+
+We need to have some metrics to be able to know if the system is working correctly or if there is a problem. 
+
+For instance, "why is the data extraction is taking so long today?" "Is it some financial institution that is having a problem?" "Is it our system?" "Our database?" 
+
+Other questions we must answer are "has the daily process already run today?" "How many API calls had a problem in the last processing?"
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:74]:
```suggestion
Other questions we must answer are "Has the daily process already run today?" "How many API calls had a problem in the last processing?"
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+Daily, this application receives huge files with investment data from other internal systems of our bank, like Investment Funds, Accounts and Shares.
+
+It is important to recognize that the mainframe environment thrives with this kind of massive processing, reading large amounts of data files and storing data on the database using specialized processes that don't execute individual SQLs. This means that we can generate a file with millions of records and execute one DB2 procedure that stores all this data very quickly.
+
+
+== Architecture to solve the problem
+
+=== New Options
+
+Banco do Brasil technology now has a private cloud orchestrated with Kubernetes, where we are able to run everything that can be built in a container. The most common languages used are Java, TypeScript, Python and Golang.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:90]:
```suggestion
Banco do Brasil technology now has a private cloud orchestrated with Kubernetes, where we are able to run everything that can be built in a container. The most common languages used are Java, TypeScript, Python and Go.
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+Quarkus was chosen by our Development Supporting Team as the official Java framework to be used. The main reasons were:
+
+* Very good development experience with fast startup time and hot reload.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:94]:
```suggestion
* Very good development experience with fast startup time and live reload.
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+* Quarkus implements the Eclipse Microprofile specification meaning that we are not locked in one specific solution, so theoretically we can change to another Microprofile implementation if needed.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:95]:
```suggestion
* Quarkus implements the Eclipse MicroProfile specification meaning that we are not locked in one specific solution, so theoretically we can change to another MicroProfile implementation if needed.
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+* The Quarkus community is very active, bugs are solved quickly and new features are released frequently.
+* Quarkus is cloud-native.
+
+Therefore, Quarkus is being massively used by us to create microservices. We have all sorts of applications that use it, the most common are simple API CRUDs, but we also have batch processes, integration with other institutions using REST APIs, integration with B3, the Brazilian stock exchange, using the FIX (Financial Information eXchange) protocol, low latency applications, and much more.
+
+NOTE: To integrate with the FIX protocol, we use https://www.quickfixj.org/[QuickFIX/J], a Java open-source solution. We created this https://github.com/felipewind/fix-trading-simulator[Fix Trading Simulator] project showing how can we use QuickFIX/J with Quarkus.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:101]:
```suggestion
NOTE: To integrate with the FIX protocol, we use https://www.quickfixj.org/[QuickFIX/J], a Java open-source solution. We created this https://github.com/felipewind/fix-trading-simulator[Fix Trading Simulator] project showing how can you use QuickFIX/J with Quarkus.
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+=== Our decision
+
+A new system was created in our company to provide integration with the other financial institutions that participate in the Brazilian Open Banking environment. It's our *Open Banking Integrator* and it was built with Quarkus. 
+
+At this point, we needed to decide how we would overcome all the project challenges. Would we stick with the mainframe, where it would be possible to communicate with our *Open Banking Integrator*, or should we try something new?
+
+As with all decisions in software architecture, we always have pros and cons. In the end, we decided to solve this problem running microservices written in Quarkus in our private cloud. The main reasons were:
+
+* It is easier to scale horizontally the process.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:111]:
```suggestion
* It is easier to scale the process horizontally.
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+But, to use one entire instance of the application to process one message at a time seems like a waste of resources. In the past our requirement was that each running pod should be able to process more than one message from the Kafka topic simultaneously. This challenge was the most exciting part of the project.
+
+The team thought that we would need to do this programatically, receiving the messages from the Kafka consumer and creating threads manually. Then, we read the https://quarkus.io/guides/kafka#blocking-processing[Quarkus Kafka guide] and discovered that it was possible to do some tuning in the worker thread pool that consumes messages. The guide says that there is more information on the SmallRye Reactive Messaging documentation. 
+
+TIP: On the https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/3.1/advanced/blocking.html[SmallRye Reactive Messaging – Handling blocking execution guide], we found that if we don't need to process our messasages in order, we actually can define the worker pool size that will consume Kafka messages. This was like magic for us! All we needed to do is to put these annotations in our Kafka Consumer:
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:147]:
```suggestion
TIP: In the https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/3.1/advanced/blocking.html[SmallRye Reactive Messaging – Handling blocking execution guide], we found that if we don't need to process our messages in order, we actually can define the worker pool size that will consume Kafka messages. This was like magic for us! All we needed to do was to put these annotations in our Kafka Consumer:
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+[source, java]
+----
+@Incoming("extraction")
+@Blocking(ordered = false, value = "extraction-pool")
+public void process(Extraction extraction) {
+  // process the extraction
+}
+----
+
+Now, we can configure the woker pool size, that means how many threads are going to process our Kafka messages simultaneously, passing this parameter:
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:158]:
```suggestion
Now, we can configure the worker pool size, that means how many threads are going to process our Kafka messages simultaneously, passing this parameter:
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+[source, properties]
+----
+smallrye.messaging.worker.extraction-pool.max-concurrency=7
+----
+
+[WARNING]
+==== 
+Each extraction demands calls to APIs that are provided in other financial institutions, sometimes this takes more time than the default amount of time defined by the Vert.x worker pool, that are 60 seconds. After 60 seconds, the application receives warnings informing us that our worker thread is blocked. It's possible to configure this with this parameter:
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:167]:
```suggestion
Each extraction demands calls to APIs that are provided by other financial institutions, sometimes this takes more time than the default amount of time defined by the Vert.x worker pool, that is 60 seconds. After 60 seconds, the application receives warnings informing us that our worker thread is blocked. It's possible to configure this with this parameter:
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+[source, properties]
+----
+# The maximum amount of time the worker thread can be blocked. Default 60S
+quarkus.vertx.max-worker-execute-time=300S
+----
+
+TIP: We created the project https://github.com/felipewind/poc-kafka-quarkus[POC Kakfa Quarkus], where you can simulate this feature of running simultaneous threads to process messages from a Kafka topic.
+
+
+=== Multiple persistence units with Hibernate
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:179]:
```suggestion
=== Multiple persistence units with Hibernate ORM
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+This feature was very important to our project. The *Open Banking Integrator* has some business data stored in an Oracle database. To improve the speed of processing, there is some information that we need to obtain directly from this database. As our application has a DB2 database, the Hibernate Multiple Persistence units feature was very helpful. More information is available on the https://quarkus.io/guides/hibernate-orm#multiple-persistence-units[Hibernate Quarkus Guide].
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:181]:
```suggestion
This feature was very important to our project. The *Open Banking Integrator* has some business data stored in an Oracle database. To improve the speed of processing, there is some information that we need to obtain directly from this database. As our application has a DB2 database, the Hibernate ORM Multiple Persistence units feature was very helpful. More information is available on the https://quarkus.io/guides/hibernate-orm#multiple-persistence-units[Hibernate Quarkus Guide].
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+
+=== Fault tolerance Retry 
+
+The APIs we execute can return errors saying that the system is temporary unavailable. In this situation, we need to wait some time and try again.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:186]:
```suggestion
The APIs we execute can return errors saying that the system is temporarily unavailable. In this situation, we need to wait some time and try again.
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+
+This is very simple when using the Retry:
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:188]:
```suggestion
This is very simple when using the `@Retry` annotation:
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+
+[source, java]
+----
+@Retry(retryOn = { ExceptionOfTheApiThatWeMustTryAgain.class }, maxRetries = 3, delay = 1000)
+public void callExternalEndpoint() {
+  // 
+}
+----
+
+We can configure the maxRetries and the delay between the retries on our properties:
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:198]:
```suggestion
We can configure the `maxRetries` and the `delay` between the retries with these properties:
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+----
+
+
+=== Synchronizing access with fault tolerance Bulkhead
+
+Our *scheduler* microservice can't run the schedule in parallel. 
+
+We deploy it on Kubernetes with the strategy type *Recreate*, meaning that all pods are terminated before a new one is created. This guarantees that only one pod at a time will be executed. So how do we avoid multiple threads from the same instance executing the same method at the same time?
+
+The answer is to use the Bulkhead fault tolerance.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:222]:
```suggestion
The answer is to use the `Bulkhead` fault tolerance annotation:
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+[source, java]
+----
+// maximum 1 concurrent requests allowed, maximum 1 requests allowed in the waiting queue
+@Bulkhead(value = 1, waitingTaskQueue = 1)
+public void processSchedule() {
+  // 
+}
+----
+
+When a request cannot be added to the waiting queue, BulkheadException will be thrown.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:233]:
```suggestion
When a request cannot be added to the waiting queue, a `BulkheadException` will be thrown.
```

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+
+
+
+== Final words about Quarkus
+
+Our organization has hundreds of Java programmers, but one problem that we see is the difficulty to hire people who already have experience with Quarkus. Although, we believe that a person who has experience with other Java frameworks can learn Quarkus very fast and, with time, we will have more professionals mastering it.
```
Comment (gsmet) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:239]:
```suggestion
Our organization has hundreds of Java programmers, but one problem that we see is the difficulty to hire people who already have experience with Quarkus. That being said, we believe that a person who has experience with other Java frameworks can learn Quarkus very fast and, with time, we will have more professionals mastering it.
```

Comment (maxandersen) [inline _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc:239]:
```suggestion
Our organization has hundreds of Java programmers, but one problem that we see is the difficulty to hire people who already have experience with Quarkus. We though believe that a person who has experience with other Java frameworks can learn Quarkus very fast and, with time, we will have more professionals mastering it.
```

**Review by gsmet — CHANGES_REQUESTED** [review]
Nice article, thanks! I suggested some improvements.

**Review by maxandersen — COMMENTED** [review]
nice one! thanks for doing it.

I just had one suggestion to make what I think is meant in the conclusion clearer (that yes, few people with existing knowledge but they can easily be taught/learn Quarkus)

File: _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc (hunk 19 above)
Comment (felipewind) [reply _posts/2022-05-25-banco-do-brasil-open-banking-user-story.adoc]: @gsmet , @maxandersen [also suggested to change this sentence](https://github.com/quarkusio/quarkusio.github.io/pull/1444#discussion_r882451294)

After accepting his suggestion, now it's like this:
> Our organization has hundreds of Java programmers, but one problem that we see is the difficulty to hire people who already have experience with Quarkus. **We though** believe that a person who has experience with other Java frameworks can learn Quarkus very fast and, with time, we will have more professionals mastering it.

Should I change to your suggestion? Both look good to me.

**Comment by felipewind** [comment]
> nice one! thanks for doing it.
> 
> I just had one suggestion to make what I think is meant in the conclusion clearer (that yes, few people with existing knowledge but they can easily be taught/learn Quarkus)

Thanks @maxandersen !

Can you help me to change this sentence and make it clearer?
> Our organization has hundreds of Java programmers, but one problem that we see is the difficulty to hire people who already have experience with Quarkus. We though believe that a person who has experience with other Java frameworks can learn Quarkus very fast and, with time, we will have more professionals mastering it.

**Review by gsmet — CHANGES_REQUESTED** [review]
Just a note that once things are settled, we need to change the date, probably to publish it on Tuesday as Monday is a slow day.

**Comment by felipewind** [comment]
> I would go with something like:
> 
> > Our organization has hundreds of Java programmers, but one problem that we see is the difficulty to hire people who already have experience with Quarkus. We though believe that a person who has experience with other Java frameworks can easily learn Quarkus - especially since Quarkus relies on existing and proven technologies (JAX-RS, CDI, Hibernate ORM, Eclipse MicroProfile...) - and, with time, we will have more professionals mastering it.

Awesome @gsmet ! I couldn't do better. I sent one commit updating the conclusion with your suggestion.

As I said before, I can't accept [this suggestion](https://github.com/quarkusio/quarkusio.github.io/pull/1444#discussion_r882451188) because @maxandersen sent first one suggestion  to the same line.

**Review by gsmet — APPROVED** [review]
I squashed everything and changed the date to Tuesday. It will be automatically published on Tuesday.

**Comment by gsmet** [comment]
@felipewind it's published and I posted a tweet about in the QuarkusIO account: https://twitter.com/QuarkusIO/status/1531580614868090880

**Pull request merged**